### DISCUSSION TITLE
Best Practices for DataTree with Multiscale Data

### DISCUSSION THREAD

**Discussion by allen-adastra** [pydata/xarray#9577] [General]
Mostly pinging @TomNicholas, I see `DataTree` is making it into the public API; exciting!

I was hoping to get your thoughts on best practices, with a concrete fusion example.

Let's consider magnetic data. We have that at three resolutions:
1. Order 10Hz from pre-shot pulse-design
2. Order 1kHz from real-time equilibrium reconstruction
3. Order 10kHz for the raw magnetic data

We also have many shots of data. Shots tend to be between 0.5 seconds and 3 seconds, which isn't a huge range, but a large enough range that we probably shouldn't just nan pad the ends and have "shot" just be a dimension along an array.

What kind of structure would you recommend with `DataTree`? Should we just have one node for each "shot"? Is there some built-in functionality to "transpose" a variable? (i.e. go from multiple arrays of "signalX" to a single array of "signalX" with a new dimension and nan padding to handle different lengths).

**Comment by TomNicholas** [comment]
> Order 10Hz from pre-shot pulse-design
Order 1kHz from real-time equilibrium reconstruction
Order 10kHz for the raw magnetic data

It's very satisfying to me to see people taking these same data management concepts and applying them to fusion :blush:

>  probably shouldn't just nan pad the ends and have "shot" just be a dimension along an array.

Yeah that's hopefully a pre-datatree anti-pattern now.

> What kind of structure would you recommend with DataTree? 

Instead of answering directly yet I'm going to use you as a guinea pig and point you to some upcoming documentation I wrote - PR #9501, the [latest build is here](https://xray--9501.org.readthedocs.build/en/9501/user-guide/hierarchical-data.html#alignment-and-coordinate-inheritance). I'm hoping reading that helps answers your question - if it doesn't then please say so.

>  Is there some built-in functionality to "transpose" a variable? (i.e. go from multiple arrays of "signalX" to a single array of "signalX" with a new dimension and nan padding to handle different lengths).

Nothing dedicated like that exists for datatree yet, but it could exist. In fact that sounds very similar to calling `xr.concat` with `join='outer'` & `fill_value=np.Nan`? (Also this is an example of https://github.com/pydata/xarray/issues/9349)

**Reply by allen-adastra** [reply]
> could

True, we can do this. Something I've noticed is that I can't have time coordinates be:
```
time (shot, time)
```
for each shot of `xr.Dataset`.

However, if I instead have `time` be a data variable and rename the `time` dimension to `time_slice` with index coordinates that is just `np.arange`, then I can do the concatenation. Wondering if this is the best practice?

**Reply by allen-adastra** [reply]
Docs look good, I'll try to get around to playing around with it; when is the release that contains `xr.DataTree` going to be out? (I can ofc install HEAD of main, but for my more productiony stuff would be good to have a proper release).

**Reply by TomNicholas** [reply]
>  I can't have time coordinates be:

Not quite sure what you mean - a more explicit example would be helpful.

> when is the release that contains xr.DataTree going to be out?

aiming for the 15th (Tuesday!)

**Reply by allen-adastra** [reply]
> 
> Not quite sure what you mean - a more explicit example would be helpful.
> 

Ended up being a mistake on my part. Needed to make sure `time` is an index coordinate.

I built and have been working with a collection of `netcdf` files, each corresponding to a shot, with order 1000 variables each. Each is about half a GB.

I've tended to find `xr.concat` across shots to be very slow once I scale up to >100 shots. I'm wondering if you have any suggestions there?

One workflow that has been suggested is instead have a single `xr.DataTree` with `shot` as a dimension instead.

**Reply by TomNicholas** [reply]
When are you doing this `xr.concat`? At opening time? Or during some analysis?

> very slow

Hard to know what's causing this without more context. 

> have a single xr.DataTree with shot as a dimension instead.

So you're suggesting instead of concatenating to create a small number of large arrays you instead store a large number of smaller arrays in a datatree?

**Reply by allen-adastra** [reply]
Right now I have a large number of datatrees, one per shot.

Would it make more sense to have a smaller number of datatrees? E.g. for groups of shots or a single datatree to rule them all?

**Reply by allen-adastra** [reply]
> When are you doing this `xr.concat`? At opening time? Or during some analysis?
> 

After opening, I grab some subset of variables and concat them.
